core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -1391,78 +1392,92 @@ class GroupCoordinator(val brokerId: Int,
     }
   }
 
-  def tryCompleteJoin(group: GroupMetadata, forceComplete: () => Boolean): Boolean = {
+  def tryCompleteJoin(
+    group: GroupMetadata,
+    generationId: Int,
+    forceComplete: () => Boolean
+  ): Boolean = {
     group.inLock {
-      if (group.hasAllMembersJoined)
+      if (generationId != group.generationId) {
+        forceComplete()
+      } else if (group.hasAllMembersJoined) {
         forceComplete()
-      else false
+      } else false
     }
   }
 
-  def onCompleteJoin(group: GroupMetadata): Unit = {
+  def onCompleteJoin(
+    group: GroupMetadata,
+    generationId: Int
+  ): Unit = {
     group.inLock {
-      val notYetRejoinedDynamicMembers = group.notYetRejoinedMembers.filterNot(_._2.isStaticMember)
-      if (notYetRejoinedDynamicMembers.nonEmpty) {
-        info(s"Group ${group.groupId} removed dynamic members " +
-          s"who haven't joined: ${notYetRejoinedDynamicMembers.keySet}")
-
-        notYetRejoinedDynamicMembers.values.foreach { failedMember =>
-          group.remove(failedMember.memberId)
-          removeHeartbeatForLeavingMember(group, failedMember.memberId)
-        }
-      }
-
-      if (group.is(Dead)) {
-        info(s"Group ${group.groupId} is dead, skipping rebalance stage")
-      } else if (!group.maybeElectNewJoinedLeader() && group.allMembers.nonEmpty) {
-        // If all members are not rejoining, we will postpone the completion
-        // of rebalance preparing stage, and send out another delayed operation
-        // until session timeout removes all the non-responsive members.
-        error(s"Group ${group.groupId} could not complete rebalance because no members rejoined")
-        rebalancePurgatory.tryCompleteElseWatch(
-          new DelayedJoin(this, group, group.rebalanceTimeoutMs),
-          Seq(GroupJoinKey(group.groupId)))
+      if (generationId != group.generationId) {
+        error(s"Received unexpected notification of join complete for ${group.groupId} " +
+          s"with an old generation $generationId while the group has ${group.generationId}.")
       } else {
-        group.initNextGeneration()
-        if (group.is(Empty)) {
-          info(s"Group ${group.groupId} with generation ${group.generationId} is now empty " +
-            s"(${Topic.GROUP_METADATA_TOPIC_NAME}-${partitionFor(group.groupId)})")
+        val notYetRejoinedDynamicMembers = group.notYetRejoinedMembers.filterNot(_._2.isStaticMember)
```

Comment:
It looks like this branch is all of the existing logic, but thought I'd check if there are any changes to look at.

Reply:
That's correct.